### PR TITLE
Add TicTacToe.sol

### PR DESCRIPTION
**Pull request by IIIIllllIIIIllllIIIIllllIIIIllllIIIIll**
This PR adds tic tac toe as a sample app.

### CODE REVIEW

**Comment by mvanderh** [comment]
@snario I think so, which reminds me I should write some docs for it.

File: packages/apps/test/ticTacToe.spec.ts [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+contract("TicTacToe", (accounts: string[]) => {
+  let game: ethers.Contract;
+
+  const stateEncoding =
+    "tuple(address[2] players, uint256 turnNum, uint256 winner, uint256[3][3] board)";
+
+  beforeEach(async () => {
+    const contract = new ethers.Contract("", TicTacToe.abi, unlockedAccount);
+    TicTacToe.link("StaticCall", StaticCall.address);
+    game = await contract.deploy(TicTacToe.binary);
```
Comment (mvanderh) [inline packages/apps/test/ticTacToe.spec.ts]: If you rebase on `develop` you can instead write 
```ts
const TicTacToe = AbstractContract.loadTruffleArtifact("TIcTacToe", {StaticCall});
game = await TicTacToe.deploy(unlockedAccount);
```

Comment (mvanderh) [reply packages/apps/test/ticTacToe.spec.ts]: Actually why are you redeploying the App before each test? I'd imagine you only need to do that once for the whole file.

**Comment by cf19drofxots** [comment]
@IIIIllllIIIIllllIIIIllllIIIIllllIIIIll rebased this to update it, and the `SharedContrats` PR is closed for now

@mvanderh are you still reviewing this?

File: contracts/apps/TicTacToe.sol [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+  struct WinClaim {
+    WinClaimType winClaimType;
+    uint256 winClaimIdx;
+  }
+
+  struct AppState {
+    address[2] players;
+    uint256 turnNum;
+    uint256 winner; // 0 => game in progress, 3 => drawn, i => player[i-1] won otherwise
```
Comment (mvanderh) [inline contracts/apps/TicTacToe.sol]: Extract these magic numbers to constants e.g. `WINNER_DRAW`.

**Comment by mvanderh** [comment]
@cf19drofxots Ya, my bad, reviewing

File: contracts/apps/TicTacToe.sol [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+    } else {
+      address loser = state.players[state.winner - 1];
+      address winner = state.players[2 - state.winner];
+
+      amounts[0] = terms.limit;
+      amounts[1] = 0;
+
+      to[0] = loser;
+      to[1] = winner;
```
Comment (mvanderh) [inline contracts/apps/TicTacToe.sol]: Wait, isn't this giving the full amount of the transfer to the `loser` and not the `winner`? 🤔

File: contracts/apps/TicTacToe.sol [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        amounts,
+        data
+      );
+    }
+
+  }
+
+  function playMove(
+    AppState state, uint256 playerId, uint256 x, uint256 y
+  ) internal pure returns (AppState) {
```
Comment (mvanderh) [inline contracts/apps/TicTacToe.sol]: Let's stick to the one-line-per-function-modifier style you used above i.e.
```solidity
function playMove(..)
  internal
  pure
  ...
```

File: contracts/apps/TicTacToe.sol [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  {
+    return state.turnNum % 2;
+  }
+
+  function applyAction(AppState state, Action action)
+    public
+    pure
+    returns (bytes)
+  {
+    AppState memory postState;
```
Comment (mvanderh) [inline contracts/apps/TicTacToe.sol]: Mm I like the name `postState`

File: contracts/apps/TicTacToe.sol [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+      require(state.board[1][winClaim.winClaimIdx] == playerId + 1, "assertWin failed");
+      require(state.board[2][winClaim.winClaimIdx] == playerId + 1, "assertWin failed");
+    } else if (winClaim.winClaimType == WinClaimType.DIAG) {
+      require(state.board[0][0] == playerId + 1, "assertWin failed");
+      require(state.board[1][1] == playerId + 1, "assertWin failed");
+      require(state.board[2][2] == playerId + 1, "assertWin failed");
+    } else if (winClaim.winClaimType == WinClaimType.CROSS_DIAG) {
+      require(state.board[2][0] == playerId + 1, "assertWin failed");
+      require(state.board[1][1] == playerId + 1, "assertWin failed");
+      require(state.board[0][2] == playerId + 1, "assertWin failed");
```
Comment (mvanderh) [inline contracts/apps/TicTacToe.sol]: Extract `playerId + 1` to a variable for readability

File: contracts/apps/TicTacToe.sol [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  enum WinClaimType {
+    COL,
+    ROW,
+    DIAG,
+    CROSS_DIAG
+  }
+
+  struct WinClaim {
+    WinClaimType winClaimType;
+    uint256 winClaimIdx;
```
Comment (mvanderh) [inline contracts/apps/TicTacToe.sol]: the "winClaim" part of `winClaimIdx` is redundant, just call it `idx` or `index` or `startIndex`.

File: contracts/apps/TicTacToe.sol [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    require(playerId == 0 || playerId == 1, "playMove: playerId not in range [0, 1]");
+
+    state.board[x][y] = playerId + 1;
+
+    return state;
+  }
+
+  function assertBoardIsFull(AppState preState) internal pure {
+    for (uint256 i=0; i<3; i++) {
+      for (uint256 j=0; j<3; j++) {
```
Comment (mvanderh) [inline contracts/apps/TicTacToe.sol]: Spacing around operators

File: contracts/apps/TicTacToe.sol (hunk 6 above)
Comment (mvanderh) [reply contracts/apps/TicTacToe.sol]: Also, a better failure message would be something like "Win claim not valid"

File: contracts/apps/TicTacToe.sol [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  function turn(AppState state)
+    public
+    pure
+    returns (uint256)
+  {
+    return state.turnNum % 2;
+  }
+
```
Comment (cf19drofxots) [inline contracts/apps/TicTacToe.sol]: Need to add `isStateTerminal`

File: contracts/apps/TicTacToe.sol (hunk 3 above)
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply contracts/apps/TicTacToe.sol]: good catch 😂

File: contracts/apps/TicTacToe.sol (hunk 9 above)
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply contracts/apps/TicTacToe.sol]: added

**Pull request merged**